File: #0041.first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {

        int size = nums.size();
        int work;
        for (int i = 0; i < size; i++) {
            while(nums[i] != i+1) {
                if (nums[i] <= 0 || nums[i] > size || nums[i] == nums[nums[i] -1]) {
                    break;
                }
                work = nums[i];
                nums[i] = nums[work-1];
                nums[work-1] = work;
            }  
        }

        for (int i = 0; i < size; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        return size + 1;  
    }
};

File: #0041.first-missing-positive.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function(nums) {
    const s = new Set(nums)
    for (let i = 1; i < 1000000; i++) {
        if (!s.has(i)) {
            return i;
        }
    }
};
